### PR TITLE
Fix memory leak in @custom_gradient closures

### LINKED ISSUE #97697: Memory leak in
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

tf 2.19

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

I believe that commit https://github.com/tensorflow/tensorflow/commit/af1889233ee4b16b1e483344a299e0536ab94b28 introduced a memory leak in the `@custom_gradient` decorator.

Specifically, [this line](https://github.com/tensorflow/tensorflow/blob/ae63b5828e6698055578a2bcb3950f7e24f92ccf/tensorflow/python/ops/custom_gradient.py#L491):

```python
result_grads = composite_tensor_gradient.replace_flat_tensors_for_gradients(
        nest.flatten(result), result_grad_components[:flat_result_len])
```

Captures the `result` object in the closure and means that this is stored in the gradient registry forever. 

Here is some code to reproduce the issue:

```python
import gc

import tensorflow as tf
from tensorflow import Tensor
from tensorflow.python.types.core import ConcreteFunction


@tf.custom_gradient
def nonsense(x):
    def grad(upstream):
      return upstream * 0.1
    return x * 2, grad

def go(outer_val):
    @tf.function
    def compute(x):
        with tf.GradientTape() as tape:
            tape.watch(x)
            y=nonsense(x)
            dy_dx = tape.gradient(y, x)
        return y, dy_dx
    print(compute(outer_val))


if __name__ == '__main__':
    for _ in range(4):
        go(tf.constant(0.1))
        gc.collect()
        concrete_function_count = len(
            [obj for obj in gc.get_objects() if isinstance(obj, ConcreteFunction)]
        )
        print("Concrete function count:", concrete_function_count)
        tensor_count = len(
            [obj for obj in gc.get_objects() if isinstance(obj, Tensor)]
        )
        print("Tensor count:", tensor_count)
```

Note: the gradient function uses a constant value here to illustrate the point, but this not representative of a real gradient function.

This outputs the following:

```
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 12
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 24
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 36
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 48
```

So we can see that, although the `tf.function` instances are being garbage collected, the `Tensor`s are not.

If we change [this line in TensorFlow](https://github.com/tensorflow/tensorflow/blob/ae63b5828e6698055578a2bcb3950f7e24f92ccf/tensorflow/python/ops/custom_gradient.py#L491) from this:

```python
result_grads = composite_tensor_gradient.replace_flat_tensors_for_gradients(
        nest.flatten(result), result_grad_components[:flat_result_len])
```

To this:

```python
result_grads = result_grad_components[:flat_result_len]
```

This resolves the issue and the `Tensor`s are garbage collected. The output in this case is:

```python
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 0
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 0
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 0
(<tf.Tensor: shape=(), dtype=float32, numpy=0.20000000298023224>, <tf.Tensor: shape=(), dtype=float32, numpy=0.10000000149011612>)
Concrete function count: 0
Tensor count: 0
```

Having said all that, gradient functions are unlikely to depend only on constant values, so even if this memory leak in TensorFlow is fixed most gradient functions will probably still leak memory with the current design of `custom_gradient`.

### Standalone code to reproduce the issue

```shell
import gc

import tensorflow as tf
from tensorflow import Tensor
from tensorflow.python.types.core import ConcreteFunction


@tf.custom_gradient
def nonsense(x):
    def grad(upstream):
      return upstream * 0.1
    return x * 2, grad

def go(outer_val):
    @tf.function
    def compute(x):
        with tf.GradientTape() as tape:
            tape.watch(x)
            y=nonsense(x)
            dy_dx = tape.gradient(y, x)
        return y, dy_dx
    print(compute(outer_val))


if __name__ == '__main__':
    for _ in range(4):
        go(tf.constant(0.1))
        gc.collect()
        concrete_function_count = len(
            [obj for obj in gc.get_objects() if isinstance(obj, ConcreteFunction)]
        )
        print("Concrete function count:", concrete_function_count)
        tensor_count = len(
            [obj for obj in gc.get_objects() if isinstance(obj, Tensor)]
        )
        print("Tensor count:", tensor_count)
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by ayushozha**
## Summary
- The `tape_grad_fn` closure in `_graph_mode_decorator` and `_eager_mode_decorator` unnecessarily captures the full `result` tensor list, preventing garbage collection of forward-pass tensors
- For the common case (non-composite tensors), `replace_flat_tensors_for_gradients` is a no-op, so capturing `result` is unnecessary
- This change avoids capturing `result` when no composite tensors are present, allowing forward-pass tensors to be GC'd immediately after the forward pass

## Test plan
- [ ] Verify existing `custom_gradient` tests pass
- [ ] Confirm memory usage does not grow when `@custom_gradient` is used repeatedly in `tf.function`

Fixes #97697

## Testing
- PASS: python -m py_compile tensorflow/python/ops/custom_gradient.py
- PASS: git diff --check HEAD^..HEAD
- NOT RUN locally: TensorFlow Bazel targets (Bazel/buildifier are unavailable in this checkout environment).

### CODE REVIEW

**Comment by ayushozha** [comment]
@rohan100jain @keerthanakadiri This fixes a **memory leak** in `@tf.custom_gradient` where gradient closures retain references to forward-pass tensors indefinitely. This causes OOM in long training runs with custom gradients. Would appreciate a review.

**Comment by keerthanakadiri** [comment]
> @rohan100jain @keerthanakadiri This fixes a **memory leak** in `@tf.custom_gradient` where gradient closures retain references to forward-pass tensors indefinitely. This causes OOM in long training runs with custom gradients. Would appreciate a review.

Hi @ayushozha, A review request has been submitted for this PR. Once approved, it will be ready for the next stage of the process. Thank you!

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Hi @Ayush10,

Thank you for contributing this fix!

The PR successfully addresses the memory leak in `@custom_gradient` closures by releasing references to the forward-pass tensors when they only contain non-composite tensors (the common case). 

We have verified this change by:
1. Running the existing custom gradient memory profiling and correctness tests, which all pass successfully.
2. Generating and running a new unit test that uses Python's `weakref` and explicit garbage collection (`gc.collect()`) to verify that the forward-pass output tensor is correctly garbage-collected when the gradient tape is discarded. The test passes successfully with this change, whereas it would fail (leaking the tensor) prior to this patch.

### Unit Test Code
Here is the unit test used to verify the fix:

```python
  def testCustomGradientNoMemoryLeak(self):
    import gc
    import weakref

    with context.eager_mode():
      @custom_gradient.custom_gradient
      def my_func(x):
        y = x * 2.0

        def grad(dy):
          return dy * 2.0

        return y, grad

      x = constant(3.0)

      def run_tape():
        with backprop.GradientTape() as tape:
          tape.watch(x)
          out = my_func(x)
          ref = weakref.ref(out)
        return tape, ref

      tape, ref = run_tape()
      del tape
      gc.collect()
      self.assertIsNone(ref())
```

Your change has been verified and is ready to be merged into the codebase. Excellent work!

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
We detected a build failure associated with this PR. It appears that a dependency is missing in the BUILD file for the `custom_gradient` target.

You added the following import in `tensorflow/python/ops/custom_gradient.py`:
```python
from tensorflow.python.framework import composite_tensor
```

However, the dependency for `composite_tensor` is not declared in the corresponding BUILD file.

**Action Required:**
Please update `tensorflow/python/ops/BUILD` to add `//tensorflow/python/framework:composite_tensor` to the dependencies of the `custom_gradient` target.

**Comment by ayushozha** [comment]
Addressed the BUILD dependency review in commit 83cc38171b2 by adding //tensorflow/python/framework:composite_tensor to the actual custom_gradient py_strict_library target. Python syntax compilation and git diff --check pass; Bazel was not available locally.

**Review by dmiltr3 — APPROVED** [review]
